docs/arcadia-users-group/20231104-testing-concepts/lesson.md

```diff
@@ -0,0 +1,662 @@
+After the dots, pytest will print summary information.
+
+**Challenge 1: Altering functions to pass all tests**
+
+Without changing the tests, alter the mean.py file from the previous section until it passes.
+When it passes, `pytest` will produce results like the following:
+
+```
+pytest
+```
+
+```
+collected 5 items
+
+test_mean.py .....
+
+========================== 5 passed in 2.68 seconds ===========================
+```
+
+<details>
+ <summary> Challenge solution </summary>
+
+There are many ways this challenge could be solved.
+One way is to check for the presence of complex numbers before calculating the mean.
+
+```
+def mean(num_list):
+    if any(isinstance(num, complex) for num in num_list):
+       return NotImplemented
+    else:
+        try:
+            return sum(num_list)/len(num_list)
+        except ZeroDivisionError :
+            return 0
+        except TypeError as original_error :
+            msg = "The algebraic mean of an non-numerical list is undefined.\
+                   Please provide a list of numbers."
+            raise TypeError(original_error.__str__() + "\n" +  msg)
+```
+
+</details>
+
+
+Note, using `pytest -v` will result in `pytest` listing which tests are executed and whether they pass or not:
+
+```
+pytest
+```
+
+```
+collected 5 items
+
+test_mean.py .....
+
+test_mean.py::test_mean_with_ints PASSED
+test_mean.py::test_mean_with_zero PASSED
+test_mean.py::test_mean_with_double PASSED
+test_mean.py::test_mean_with_long PASSED
+test_mean.py::test_mean_with_complex PASSED
+
+========================== 5 passed in 2.57 seconds ===========================
+```
+
+As we write more code, we would write more tests, and pytest would produce more dots.
+
+### Parametrized tests
+
+Often, different unit tests of the same "unit" have similar logic, but are applied to different inputs; this is the case with our `mean()` test suite above.
+`pytest` offers a convenient mechanism to write just a single test and apply it to multiple inputs, called test parameterization.
+This way, adding a new test case is as simple as defining one more input and expected output.
+
+To parametrize a test, we "decorate" it with its expected inputs and outputs, and `pytest` will expand the test to run as many times as needed to check all inputs.
+In the `pytest` output, this will look exactly like running multiple tests that have different functions.
+
+<details>
+<summary> Python decorators </summary>
+"Decorators" are a Python way to enhance or wrap a function with additional behavior.
+They are added just before a function definition, using a name that starts with `@`, and can take arguments separate from the function arguments, for example:
+
+<pre><code>
+@log_to_file("output.txt")
+def mean(num_list):
+  ...
+</code></pre>
+</details>
```

Comment:
This might be too much but I think it would help to demystify decorators to show explicitly what "wrapping" means:

```suggestion
This code is equivalent to wrapping or "decorating" the `mean` function with the function `log_to_file("output.txt")`:
<pre><code>
def _mean(num_list):
    ...

mean = log_to_file("output.txt")(_mean)
</code></pre>
</details>
The ampersand-based "decorator" syntax is just a clearer and more readable way of wrapping a function with another function. In this example, that other function is in fact `log_to_file("output.xt")`. How is this possible? `log_to_file` is a function that takes a filename as an input and _returns another function_ that itself takes a function as an argument and then returns a new "wrapped" function. 
```